### ISSUE TITLE
unexpected file permission error in container

### ISSUE THREAD

**Issue by astraw** [docker/docker#783]
I've narrowed down a problem in an originally more involved setup. Consider the following Dockerfile:

    # Dockerfile
    FROM      ubuntu:12.10

    RUN apt-get install -y puppetmaster sudo

    RUN rm -rf /etc/puppet
    ADD puppet-config /etc/puppet
    RUN chown -R puppet.puppet /etc/puppet
    RUN chmod 755 /etc/puppet

When run with the following:

    # make a dummy directory
    mkdir puppet-config
    echo "hi" >puppet-config/hello.txt

    docker build -t dockbug .

    echo "note the directory is owned by puppet with full read/write/execute privs"
    docker run dockbug ls -al /etc/puppet

    echo "but we get a permission error here"
    docker run dockbug sudo -u puppet ls -al /etc/puppet

I see an unexpected permission error in the final command. This is with Docker 0.3.4 from the PPA on Ubuntu 13.04 with kernel 3.8.0-19-generic. Interestingly, if I remove the like "RUN rm -rf /etc/puppet" from the Dockerfile, I no longer see the permission error.

**Comment by jerng** [comment]
Ubuntu, Docker, PHP/nginx:

Symptom: a COPYed directory was fully manipulable in the shell, according to its explicit permissions. However, PHP5-FPM would run into (permission denied) while opening files no matter what ownership and bitmask they had. 

Workaround: using VOLUME and (run -v) removed the paradox, so I'm guessing this is related.

**Comment by fredlf** [comment]
@SvenDowideit This seems common and long-lived enough that we should document it. I'm going to make a new "Known Issues" page, linked from Release Notes, in a separate PR. We'll add this stuff there.

**Comment by WiR3D-kNiGhT** [comment]
@fredlf common and long lived yes
well known and easily fixable no. 

Known issues is a start, but it could use some linking from the VOLUME sections on the relevant pages.

**Comment by WiR3D-kNiGhT** [comment]
@fredlf maybe also listing common symptoms? 
php5-fpm permission issues is a big one.

**Comment by jdef** [comment]
Looks like this will finally be solvable via #9684. Great!

http://aufs.sourceforge.net/aufs3/man.html

> By default (`nodirperm1`), aufs repects the directory permission bits on all branches equally, which means if the permission bits for a directory on a lower readonly branch prohibits you to read, then you cannot read even if you run `chmod a+rx` (and aufs copies it up). With this option (`dirperm1`), the behaviour changes and aufs checks the permission bits of the directory on the topmost branch and the permission bits on all lower branches are ignored. In other words, you read a directory even if the lower readonly branch fs prohibits it by its permission bits.
> This feature may invite a security risk similar to the world writable upper branch. As this case, `dirperm1` option will produce a warning too.

**Comment by jberkus** [comment]
What release is this likely to be fixed in?

**Comment by cpuguy83** [comment]
@jberkus This is specifically an issue with AUFS and not Docker itself.

**Comment by jberkus** [comment]
jdef's comment above seems to indicate that it's fixable at the Docker level.  No?

**Comment by cpuguy83** [comment]
So I'm not able to repro this, even with the OP's script.
Can someone provide a way to trigger this?

**Comment by j0hnsmith** [comment]
@cpuguy83 You should be able to reproduce with this dockerfile that uses a hack to get around the problem

https://github.com/orchardup/docker-postgresql/blob/master/Dockerfile

**Comment by estesp** [comment]
PR #10390 will allow using the `dirperm1` option to AUFS to fix the problem--however, that will require Ubuntu 14.10 or later, or a custom kernel with the more recent AUFS version(s) that support the option.

**Comment by jberkus** [comment]
This seems like a pretty severe and pervasive bug, given the sheer number of other issues which have been forwarded to it.  Waiting for Linux 3.19 to be mainstream is going to be pretty painful; it means Docker will keep getting slews of bug reports on related issues for 2 years or more.  At the least, warnings about this will need to be all over the documentation.

Does this affect RHEL/Centos/Fedora at all?

**Comment by jfrazelle** [comment]
It the no 1 known issue listed in the docs :)

**Comment by jberkus** [comment]
jfrazelle, I mean we'll need to add it to the ADD and COPY docs for Dockerfiles, among other things. 

It's also kind of unclear at this point how to reliably build Ubuntu/Mint-based images which have daemons running as non-root who own their own files, reliably at scale.  That's why I'm asking about Fedora; maybe the answer is that users should use Fedora, it if doesn't have this issue.

**Comment by thaJeztah** [comment]
I had some code lying around to reproduce this, copied it in a gist;
https://gist.github.com/thaJeztah/8aa9e0760e63262b8aea

**Comment by feiyang21687** [comment]
docker daemon terminated, below is ```dmesg``` info

```
[1169979.957724] init: docker main process (16734) terminated with status 2
[1169979.957779] init: docker main process ended, respawning
[1169982.778778] type=1400 audit(1437514305.726:37): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="docker-default" pid=18897 comm="apparmor_parser"
[1169983.063454] aufs au_opts_parse:1155:docker[18893]: unknown option dirperm1
```

docker info

```
Containers: 4
Images: 51
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 59
 Dirperm1 Supported: false
Execution Driver: native-0.2
Kernel Version: 3.13.0-32-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 2
Total Memory: 1.954 GiB
Name: usertagCluster-ecs.s2.small-9d6c4780d25dddbd764970a37b36788b
ID: YDPA:3IZC:OG4A:KAX3:DESX:OOEQ:DU32:D2YQ:6S5G:UYCA:NWEQ:CNO5
WARNING: No swap limit support
```

docker version

```
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2
OS/Arch (server): linux/amd64
```

uname -a

```
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2
OS/Arch (server): linux/amd64
root@usertagCluster-ecs:/var/log# uname -a
Linux usertagCluster-ecs.s2.small-9d6c4780d25dddbd764970a37b36788b 3.13.0-32-generic #57-Ubuntu SMP Tue Jul 15 03:51:08 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by jberkus** [comment]
Feiyang,

I am unclear on what your crash report has to do with the named bug.  Also, this bug is closed.

**Comment by feiyang21687** [comment]
@jberkus 

We haven't found the reason of daemon restart. We just get some kernel logs in ```dmesg```, and the time can be matched.

```
Jul 22 05:31:46 usertagCluster-ecs kernel: [1169983.063454] aufs au_opts_parse:1155:docker[18893]: unknown option dirperm1
```

```
root     18878     1  0 05:31 ?        00:01:06 /usr/bin/docker -d
```

so we suspect the ```unknown option dirperm1``` is the cause. And we google it, then found this issue. 

We need help, we want to know whether or not ```unknown option dirperm1``` could cause daemon terminate?

**Comment by alercunha** [comment]
I am experiencing this same issue sing Devicemapper, I don't think it's related to AUFS. The work around to this is doing a rm -fr * at the folder level.

**Comment by fijimunkii** [comment]
I'm unfortunately seeing this in system logs and the symptoms are no ssh or http(s) access, not sure what happened. High cpu usage visible from CloudWatch

`login: [   70.068342] aufs au_opts_parse:1155:docker[1041]: unknown option dirperm1`

**Comment by vielmetti** [comment]
Ran into this today in the https://github.com/vpicavet/docker-pggis system, trying to get PostGIS running in a container.

**Comment by thaJeztah** [comment]
@fijimunkii @vielmetti fixing this depends on the version of aufs being present on the docker host, and requires a dynamically linked binary (older versions of docker are statically linked). We now offer dynamically linked binaries through our apt and yum repos (see blog post: https://blog.docker.com/2015/07/new-apt-and-yum-repos/), and those are installed by default by the installation script, and we're updating our install docs (see https://github.com/docker/docker/pull/16662)

To check if your system supports the dirperm1 option, check the output of `docker info`:
```
...
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 7
 Dirperm1 Supported: true
```

**Comment by gerrywastaken** [comment]
> It the no 1 known issue listed in the docs :)
@jfrazelle I'm having a tough time finding that page or any known issues page. The closest I've been able to find is a know issues page for Kitematic:
https://www.google.com.au/#q=site:docker.com+%22known+issues%22

I know I'm just missing an obvious page, but do you have a link?

**Comment by vielmetti** [comment]
Thanks @thaJeztah . My environment is on a Mac using docker-machine to bring up a VM under virtualbox. The bit of relevant info to the file system is

```
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 263
 Dirperm1 Supported: true
```

which looks right. The full "docker info" output is here: 

https://gist.github.com/8ef2de84a52c266b7435

**Comment by thaJeztah** [comment]
@vielmetti if you're using bind-mounted volumes for your container and have permission problems, chances are that's related to https://github.com/boot2docker/boot2docker/issues/581 (or  https://github.com/kitematic/kitematic/issues/351)

**Comment by fijimunkii** [comment]
Thanks @thaJeztah! I tried a few ways of enabling dirperm1 support, but in the end switched to Overlay FS with [this guide](https://gist.github.com/parente/025dcb2b9400a12d1a9f)

**Comment by jberkus** [comment]
Just for paper trail:

As of:
* Docker 1.8
* Ubuntu, Kernel 3.13.0-55

... now getting the "can't change permissions on directories" issue on all of my images instead of just some of them.

**Comment by thaJeztah** [comment]
@jberkus could you open a new issue? please provide the output of `docker version` and `docker info` as well when reporting, and if possible the exact steps to reproduce.

**Comment by jberkus** [comment]
Sure, will do.

**Comment by shykes** [comment]
This issue is still linked from https://docs.docker.com/engine/reference/builder/ in spite of being closed. Is that normal?

**Comment by thaJeztah** [comment]
@shykes it's only fixed for distro's having a recent version of aufs; I should check, but possibly it's not fixed on Ubuntu 14.04. If it *is* fixed, I think we should remove the "known issue"

**Comment by AkihiroSuda** [comment]
#20240 is similar, but Dirperm1 is enabled.

**Comment by antoineco** [comment]
Running into this issue with Docker for OS X beta (Docker 1.11.0-beta8) while using ONBUILD instructions. It's quite easy to reproduce:

```dockerfile
FROM debian
RUN addgroup --gid 9999 app
RUN adduser --uid 9999 --gid 9999 --system --disabled-login app
RUN apt update && apt install -y nginx-light
RUN chown app /var/log/nginx
```

```
$ docker build -t issue783 .
$ docker run -it --rm issue783 bash -il
```

```
(root)# ls -dn /var/log/nginx/
drwxr-x--- 2 9999 4 4096 Feb 10 21:54 /var/log/nginx/
(root)# runuser -u app -- bash -il
(app)$ id
uid=9999(app) gid=9999(app) groups=9999(app)
(app)$ touch /var/log/nginx/test
touch: cannot touch '/var/log/nginx/test': Permission denied
```

It works if you move the directory and try again:

```
(root)# mv /var/log/nginx/ /var/log/nginx_new/
(root)# runuser -u app -- bash -il
(app)$ touch /var/log/nginx_new/test
(app)$ ls -l /var/log/nginx_new/
total 0
-rw-r--r-- 1 app app 0 Apr 21 13:02 test
```

**Comment by thaJeztah** [comment]
@antoineco I was able to reproduce on Docker for OS X, switching to overlay resolved it, so it's an aufs issue indeed. Can you report this issue through beta-feedback@docker.com ?

**Comment by thaJeztah** [comment]
I opened an issue in the internal issue tracker as well (number `#2693`)

**Comment by jberkus** [comment]
FWIW, I'm on Fedora 23 now, and haven't had the issue since.  So it is getting resolved upstream.

**Comment by thaJeztah** [comment]
@jberkus thanks for the info. I was able to reproduce it on Ubuntu 14.04 on kernel 3.19.something. Would be nice to know which version of aufs its solved in, possibly the Docker Mac team can bump the version to include the patch that's needed

**Comment by stuartnelson3** [comment]
I can confirm encountering this issue on the the docker os x beta-15 (current version). I have switched back to docker-machine in the meantime.

**Comment by kuznero** [comment]
`rm /etc/puppet/*` didn't work for me, instead I found that deleting individual files on the same layer works perfectly well:

```
find /etc/pappet -type f | xargs -L1 rm -f
```

**Comment by glasser** [comment]
We are seeing a very similar issue with Overlay.

Our current guess is that it is caused by the bug fixed in kernel 4.4.6 by this commit: https://lkml.org/lkml/2016/1/31/82

We have not yet managed to test with 4.4.6.

**Comment by LittleControl** [comment]
> `rm /etc/puppet/*` didn't work for me, instead I found that deleting individual files on the same layer works perfectly well:
> 
> ```shell
> find /etc/pappet -type f | xargs -L1 rm -f
> ```
> 
> Here is link to my [Gist](https://gist.github.com/kuznero/dcc2c427d76a1e5f7e2190a4f5e23b7d)

Thanks, it works for me!
